database/yugabytedb/yugabytedb.go

```diff
@@ -363,50 +405,55 @@ func (c *YugabyteDB) ensureLockTable() error {
 	return nil
 }
 
-func ExecuteInTx(ctx context.Context, db *sql.DB, opts *sql.TxOptions, fn func(tx *sql.Tx) error) (err error) {
-	// It looks like currently, YugabyteDB doesn't return an error on savepoint release.
-	// So, we have to retry the whole transaction.
-	// TODO: do more research on savepoints support, see https://github.com/yugabyte/yugabyte-db/issues/9219
+func (c *YugabyteDB) doTxWithRetry(
+	ctx context.Context,
+	txOpts *sql.TxOptions,
+	fn func(tx *sql.Tx) error,
+) error {
+	backOff := c.newBackoff(ctx)
 
-	const maxRetries = 50
-	retryCount := 0
-
-	for {
-		if retryCount-1 >= maxRetries {
-			return fmt.Errorf("%w: reached %d attempts", ErrMaxRetriesExceeded, retryCount-1)
-		}
-
-		tx, err := db.BeginTx(ctx, opts)
+	return backoff.Retry(func() error {
+		tx, err := c.db.BeginTx(ctx, txOpts)
 		if err != nil {
-			return err
+			return backoff.Permanent(err)
 		}
 
-		if err := fn(tx); err != nil {
-			_ = tx.Rollback()
-
-			if !errIsRetryable(err) {
-				return err
-			}
+		defer tx.Rollback()
 
-			retryCount++
+		if err := fn(tx); err != nil && !errIsRetryable(err) {
+			return backoff.Permanent(err)
+		} else if err != nil {
+			return err
+		}
 
-			continue
+		if err := tx.Commit(); err != nil && !errIsRetryable(err) {
+			return backoff.Permanent(err)
+		} else if err != nil {
+			return err
 		}
 
-		if err := tx.Commit(); err != nil {
-			_ = tx.Rollback()
+		return nil
+	}, backOff)
+}
 
-			if !errIsRetryable(err) {
-				return err
-			}
+func (c *YugabyteDB) newBackoff(ctx context.Context) backoff.BackOff {
+	if ctx == nil {
+		ctx = context.Background()
+	}
 
-			retryCount++
+	retrier := backoff.WithMaxRetries(backoff.WithContext(&backoff.ExponentialBackOff{
+		InitialInterval:     backoff.DefaultInitialInterval,
+		RandomizationFactor: backoff.DefaultRandomizationFactor,
+		Multiplier:          backoff.DefaultMultiplier,
+		MaxInterval:         c.config.RetryMaxInterval,
+		MaxElapsedTime:      c.config.RetryMaxElapsedTime,
+		Stop:                backoff.Stop,
+		Clock:               backoff.SystemClock,
+	}, ctx), uint64(c.config.RetryMaxRetries))
 
-			continue
-		}
+	retrier.Reset()
```

Comment:
Do we need to reset `retrier` after it's been created?

Reply:
I think yes. It sets `currentInterval` and `startTime` which are empty. Doc comments says ` Reset must be called before using b`.

Reply:
Ahh, ok. Thanks for the explanation!
Our other usage used [NewExponentialBackOff](https://pkg.go.dev/github.com/cenkalti/backoff/v4#NewExponentialBackOff) which calls `Reset()` internally.